vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py

```diff
@@ -259,6 +266,93 @@ def shutdown(self):
             self.connector_worker.shutdown()
 
 
+class ReqsNeedSendTracker:
+
+    @dataclass
+    class RequestTimer:
+        """Timer for requests that need to be sent for remote decode."""
+        expiry_time: float
+        """Expiry time to avoid stranded KV blocks that are never fetched."""
+        consumer_count: int
+        """Consumer notification count - with heterogeneous TP, P must wait
+        for all assigned D TP workers to finish reading before safely freeing
+        the blocks."""
+
+    def __init__(self):
+        self._reqs_need_send: dict[ReqId,
+                                   ReqsNeedSendTracker.RequestTimer] = {}
+        self._timeout = envs.VLLM_NIXL_ABORT_REQUEST_TIMEOUT
+
+    def start_timer(self, req_id):
+        self._reqs_need_send[req_id] = self.RequestTimer(
+            time.monotonic() + self._timeout, 0)
+
+    def delete_timer(self, req_id):
+        if req_id not in self._reqs_need_send:
+            return
+        logger.debug("Deleting KV transfer timeout for request %s", req_id)
+        del self._reqs_need_send[req_id]
+
+    def _process_finished_notifs(self, finished_notifs: set[str]) -> set[str]:
+        """Process notifications from D and track consumer completion.
+
+        The notification strings are in format "req_id:tp_ratio".
+
+        Return request IDs that have completed sending to all consumers, to be
+        used by the scheduler via KVConnectorOutput.finished_sending.
+        """
+        finished_sending: set[str] = set()
+        for notif in (finished_notifs or ()):
+            try:
+                req_id, tp_ratio = notif.rsplit(":", 1)
+            except (ValueError, TypeError) as e:
+                raise ValueError(f"Invalid notification: {notif}") from e
+
+            # Sent notifications received after we already timed out
```

Comment:
Corner case ... we time out (or abort?) and then later receive notifications? Is it possible? We have no test to cover this